networkx/algorithms/community/label_propagation.py

```diff
@@ -0,0 +1,149 @@
+    labeling = _uniquely_label(G)
+    nrounds = 1
+    while not _labeling_complete(labeling, G):
+        nrounds += 1
+        _update_labels(labeling, coloring, G)
+        
+    for label in set(labeling.values()):
+        yield set((x for x in labeling if labeling[x] == label))
+    
+    
+def _break_color_tie(current, labels):
+    """Uses Prec-Max tie-breaking to break the ties, as laid out in:
+       'Community Detection via Semi-Synchronous Label Propagation Algorithms'
+       Cordasco and Gargano, 2011
+
+       If the labels set specified is empty than the current label is returned.
+
+       Specified set of labels is assummed to be one of integers.
+    """
+    if len(labels) == 0 or current in labels:
+        new_label = current
+    else:
+        new_label = max(labels)
+    return new_label
+
+def _calculate_label_frequencies(node, labeling, G):
+    """Counts up the labels of the neighbors of the specified node. Returns a
+       dictionary from the label to the frequency.
+    """
+    return dict(Counter(labeling[q] for q in G[node]))
+
+def _color_network(G):
+    """Colors the network so that neighboring nodes all have distinct colors.
+       Returns a dict of set of nodes and also a lookup of nodes to colors, in 
+       a tuple in that order.
+    """
+    coloring = dict() # color => set(node)
+    lookup = dict() # node => color
+    finally_colored_nodes = set()
+    
+    for n in G.nodes():
+        if n not in lookup:
+            color = 0                 
+        else:
+            color = lookup[n]
+        lookup[n] = color
+        finally_colored_nodes.add(n)
+        if color not in coloring: 
+            coloring[color] = set()
+        coloring[color].add(n)
+        for q in G.neighbors(n):
+            if q not in finally_colored_nodes: 
+                lookup[q] = lookup[n] + 1
+    
+    return coloring
+
+def _form_communities(labeling, G):
+    """Determines the communities from the labels of the network, returning a
+       dict of sets of nodes.
+    """
+    return {k: set(v) 
+            for k, v in groupby(sorted(G, key=labeling.get), key=labeling.get)}
+
+def _labeling_complete(labeling, G):
+    """Determines whether or not LPA is done. It is complete when all nodes 
+       have a label that is in the set of highest frequency labels amongst its
+       neighbors.
+
+       Nodes with no neighbors are themselves a community and are therefore
+       labeled, hence the immediate if statement in the for loop.
+    """    
+    return all(labeling[v] in 
+                    _select_labels_of_highest_frequency(v, labeling, G) 
+                        for v in G if len(G[v]) > 0)
+
+def _select_labels_of_highest_frequency(node, labeling, G):
+    """Finds all labels of maximum frequency. Specified freqs must be a mapping
+       from label to frequency of that label.
+
+       Returns a set.
+    """
+    freqs = _calculate_label_frequencies(node, labeling, G)
```

Comment:
The function `_calculate_label_frequencies` is only used once here. Should probably just put the one line code here with a comment above it: `# Compute the frequencies of each node label in the graph.`